Scanner object considering multiple coils

Stockless opened the pull request:
Generation of a flexible and extensible Scanner object, to controll Hardware Limits, Gradient Behaviour, RF behaviour and magnetic field inhomegeneities.
Based in the pull request #454 and solving the issue #150 , this PR decouple the coils definition from the Phantom structure.

Clone this branch to test it using Julia Pkg:
```julia-repl
pkg> add https://github.com/JuliaHealth/KomaMRI.jl:KomaMRIBase#scanner-develop
pkg> add https://github.com/JuliaHealth/KomaMRI.jl:KomaMRICore#scanner-develop
pkg> add KomaMRI
```

cncastillo committed:
Coil-sens on Scanner

cncastillo committed:
Coil-sens on Scanner

pvillacorta committed:
First commit

pvillacorta committed:
Commit of the following:
- Solve MotionList(rot) bug
- Pass Docs CI
- Add useful constructors to `Motion`

pvillacorta committed:
Commit of the following:
- Finish lit-1-phantom.jl
- New lit-3-phantom-format.jl
- Upload svg files for phantom file tree
- Update cross references

pvillacorta committed:
Solve bugs:
- Typo in lit-1-phantom.jl
- Typo in ph-timespan-type svg filenames
- Fix light mode for ph-timespan-type

pvillacorta committed:
Commit of the following:
- Change phantom-format file to pure markdown
- Try `execute=true` option for Literate.markdown function
- Change `plot_phantom_map` example to display T1
- Remove buildkite test to avoid overloading

pvillacorta committed:
Commit with the following:
- Remove `execute=true`
- Update svg files for file format section
- Display outputs in phantom section

pvillacorta committed:
Correct docstring

cncastillo committed:
Remove Kaleido_jll compat entry

This is meant to fix the issues with Apple Silicon Mac OS.

Commit:
CompatHelper: bump compat for oneAPI in [weakdeps] to 2 for package KomaMRICore, (keep existing compat)

rkierulf committed:
Initial commit

rkierulf committed:
synchronize immediately after kernels finish

rkierulf committed:
add -g2 to AMD and CUDA tests to see error

rkierulf committed:
change synchronization back

rkierulf committed:
remove -g2

rkierulf committed:
fix sequence indexing issue

rkierulf committed:
Adapt to KernelAbstractions 0.9.34 + clean up code

Stockless committed:
Delete unintended and unnecesary elements

Stockless committed:
Delete unintended and unnecesary elements

rkierulf committed:
Use shuffle_down intrinsic for faster reduction

Stockless committed:
coil_sens removal from Phantom structure

cncastillo commented on KomaMRIBase/src/sequences/PulseDesigner.jl:
I suggest to add a `getproperty(sys::Scanner, key::Symbol)` that calls `sys.limits.symbol` if the symbol is a field of `HardwareLimits`.

Stockless replied:
Solved in [4dbf3d1](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/4dbf3d10c5f336d0ca624c5e2454adc2e22ae511)

cncastillo commented on docs/Project.toml:
```diff
@@ -1,7 +1,10 @@
 [deps]
 Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
+Interpolations = "a98d9a8b-a2ab-59e6-89dd-64a1c18fca59"
 KomaMRI = "6a340f8b-2cdf-4c04-99be-4953d9b66d0a"
 Literate = "98b081ad-f1c9-55d3-8b20-4c87d4299306"
+MAT = "23992714-dd62-5051-b70f-ba57cb901cac"
+MRIReco = "bdf86e05-2d2b-5731-a332-f3fe1f9e047f"
```
I don't think we need this new parameters, KomaMRI re-exports MRIReco, and Interpolations and MAT are not needed.

Stockless replied:
Solved in [4dbf3d1](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/4dbf3d10c5f336d0ca624c5e2454adc2e22ae511)

cncastillo commented on examples/3.tutorials/lit-08-CoilsCircle.jl:
Remove.

Stockless replied:
Solved in [4dbf3d1](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/4dbf3d10c5f336d0ca624c5e2454adc2e22ae511)

rkierulf committed:
Fix typo

Stockless committed:
remove unnecesary example file, remove unnecesary packages call and  add a getprperty to the Scanner.limits.symbol values.

Stockless committed:
remove unnecesary example file, remove unnecesary packages call and  add a getprperty to the Scanner.limits.symbol values.

cncastillo commented on KomaMRICore/src/simulation/SimMethods/Bloch/BlochGPU.jl:
```diff
@@ -159,7 +160,8 @@ function run_spin_precession!(
 
-        sig .= transpose(sum(pre.Mxy; dims=1))
+        acquire_signal!(sig, sys.rf_coils, pre.Mxy)
+        #sig .= transpose(sum(pre.Mxy; dims=1))
```
```julia
#sig .= transpose(sum(pre.Mxy; dims=1))
```

You can probably remove this line

Stockless replied:
Solved in [821f765](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/821f765589a46e9d83da208419e65c1fab958f6b)

cncastillo commented on KomaMRICore/src/simulation/SimMethods/SimulationMethod.jl:
```diff
@@ -10,7 +10,9 @@ include("Magnetization.jl")
 ) where {T<:Real}
-    return (sum(seq.ADC.N), 1) #Nt x Ncoils, This should consider the coil info from sys
+    # Determine the number of coils
+    n_coils = size(sys.rf_coils.coil_sens, 2)
```
This will only work for ArbitraryRF no? Better to have a `get_n_coils(::RFCoils)`.

Stockless replied:
Do I also remove the obj:Phantom{T} argument in this function? it's not being used anymore

cncastillo replied:
I didn't notice, yes please remove this of all places that it was used.

cncastillo replied:
Why is does that change not appear in the diff?

cncastillo replied:
Don't remove the `Phantom{T}` if it was like that before.

Stockless replied:
Solved in [821f765](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/821f765589a46e9d83da208419e65c1fab958f6b)

cncastillo commented on Project.toml:
```diff
@@ -20,9 +20,6 @@ AssetRegistry = "0.1"
 Blink = "0.12"
 FFTW = "1.0"
 Interact = "0.10"
-KomaMRICore = "0.9"
-KomaMRIFiles = "0.9"
-KomaMRIPlots = "0.9"
```
Don't remove these!

Stockless replied:
Solved in [821f765](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/821f765589a46e9d83da208419e65c1fab958f6b)

cncastillo commented on examples/2.phantoms/sphere_fields.mat:
Remove this!

Stockless replied:
Solved in [71c20e3](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/71c20e3b8b3b11cd1e4e718e4955dc9fd01d12bd)

Stockless committed:
add a get_n_coils function, re-add KomaMRI subpackages to KomaMRI [compact] and another small changes

Stockless committed:
add a get_n_coils function, re-add KomaMRI subpackages to KomaMRI [compact] and another small changes

Stockless committed:
remove sphere_fields.mat

Stockless commented on KomaMRIBase/src/datatypes/Scanner.jl:
```diff
@@ -108,4 +108,12 @@ function getproperty(sys::Scanner, key::Symbol)
 end
 
-export ArbitraryRFCoils, RFCoilsSensDefinedAtPhantomPositions, UniformRFCoils, acquire_signal!, HardwareLimits, LinearXYZGradients
+function get_n_coils(rf_coils::UniformRFCoils)
+    return 1
+end
+
+function get_n_coils(rf_coils::RFCoils)
+    return size(rf_coils.coil_sens, 2)
+end
```
I did this to verify the function:
```julia
julia> using KomaMRI
julia> obj = brain_phantom2D()
julia> sys = Scanner()
julia> get_n_coils(sys.rf_coils)
1
julia> coil_sens = [1.0 1.0]
julia> sys.rf_coils = RFCoilsSensDefinedAtPhantomPositions(complex.(coil_sens))
RFCoilsSensDefinedAtPhantomPositions{Float64}(ComplexF64[1.0 + 0.0im 1.0 + 0.0im])
julia> get_n_coils(sys.rf_coils)
2
julia> sys.rf_coils = ArbitraryRFCoils(obj.x, obj.y, obj.z, complex.(coil_sens), complex.(coil_sens))
julia> get_n_coils(sys.rf_coils)
2
julia> get_n_coils(sys.gradients)
ERROR: MethodError: no method matching get_n_coils(::LinearXYZGradients{Float64})
The function `get_n_coils` exists, but no method is defined for this combination of argument types.
```
Also checked that the added example runs the simulation correctly with both `UniformRFCoils` and `RFCoilsSensDefinedAtPhantomPositions`

cncastillo replied:
Please change it to match what I said.

Stockless replied:
Solved at [766cb48](https://github.com/JuliaHealth/KomaMRI.jl/pull/548/commits/766cb485419fc5d30dd0c4e1a937880aeb745c70)

Stockless committed:
redefined get_n_coils function: 1 for abstract type and size of coil sens for RFCoilSensDefinedAtPhantomPositions

cncastillo committed:
Merge branch 'master' into scanner-develop

Stockless committed:
Correction of Scanner argument at run_spin_precession functions

Stockless committed:
Correction of Scanner argument at run_spin_precession functions

rkierulf committed:
fix warp reduction

rkierulf committed:
change default precession groupsize to 256

rkierulf committed:
Is warp reduction actually faster?

rkierulf committed:
Continue using warp reduction

rkierulf committed:
Update docs

rkierulf committed:
Debug AMDGPU error

rkierulf committed:
Undo debug statements skip benchmarks skip ci

rkierulf committed:
Temporarily add HIP_LAUNCH_BLOCKING

Stockless committed:
Add "How to test contributions" in "How to contribute to Koma" documentation

cncastillo committed:
Modified `center-images.css` to avoid centering badges.

Stockless committed:
Instructions less verbose

rkierulf committed:
Print matrix values skip benchmarks

rkierulf committed:
Remove suppress

rkierulf committed:
Remove debug statements skip benchmarks skip ci

Stockless committed:
adding specific dispatch for CPU

Stockless committed:
Using intended acquire_signal for BlochCPU (not working on multiple coils yet)

Stockless committed:
Fixed BlochCPU with multiple coils

Stockless committed:
How to add a new test

Stockless committed:
add setpropery to directly define hardware limitations on the scanner structure

cncastillo commented on KomaMRIBase/src/datatypes/Scanner.jl:
```diff
@@ -108,6 +108,14 @@ function getproperty(sys::Scanner, key::Symbol)
 
+function Base.setproperty!(sys::Scanner, key::Symbol, value)
+    if key in fieldnames(HardwareLimits)
+        setfield!(sys.limits, key, convert(Float64, value))  # Convert value to the correct type
```
Don't convert inside. This is going to generate type stability problems.

Stockless committed:
remove type conversion and using HardwareLimitations as Float

Stockless committed:
Fix sys.Smax to be Float in KomaMRICore/runtest

pvillacorta committed:
Merge branch `master` into `docs-phantom`

pvillacorta committed:
Adapt current docs to the new v0.9.1 `Phantom`

pvillacorta committed:
Update image source in Phantom format documentation

pvillacorta committed:
Write the motion section and minor changes in other related sections

pvillacorta committed:
Fix typos in lit-2-motion.jl

pvillacorta committed:
Hide lines in lit-2-motion.jl and plot entire phantom

pvillacorta committed:
Refactor displacement functions to center coordinates before rotation calculations

pvillacorta committed:
Fix relative paths in lit-2-motion.jl for correct navigation

pvillacorta committed:
Definitely fix relative paths in lit-2-motion.jl

pvillacorta committed:
Update references and adjust plot heights in lit-2-motion.jl and 2-koma-base.md

pvillacorta committed:
Fix erroneous semicolons in lit-2-motion.jl

pvillacorta committed:
Update rotation tests for KomaMRIBase

pvillacorta committed:
Fix cross-reference links to `KomaMRIBase.TimeCurve`

pvillacorta committed:
Fix comment line

pvillacorta committed:
Add section on Phantom File I/O and clarify .phantom format description

Stockless committed:
restrict PlutoPlotly to version 0.5

Stockless committed:
trivial change to test documentation

Stockless committed:
bump KomaMRIPlots to version 0.9.2

Commit:
CompatHelper: bump compat for PlutoPlotly in [weakdeps] to 0.6 for package KomaMRIPlots, (keep existing compat)

Stockless committed:
Notebook Update

Stockless committed:
hide reproducibility raw md text

Stockless committed:
Unsafe_blink macro implementation

pvillacorta committed:
Extend "Create your own Phantom" section

pvillacorta committed:
Reduce plots height and improve introductive paragraph.

pvillacorta committed:
Remove unnecessary example output

dependabot[bot] committed:
Bump julia-actions/julia-format from 3 to 4

Bumps [julia-actions/julia-format](https://github.com/julia-actions/julia-format) from 3 to 4.
- [Release notes](https://github.com/julia-actions/julia-format/releases)
- [Commits](https://github.com/julia-actions/julia-format/compare/v3...v4)

---
updated-dependencies:
- dependency-name: julia-actions/julia-format
  dependency-version: '4'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

pvillacorta committed:
Fix formatting of the '.phantom' file extension in documentation

aTrotier commented:
The branch works on my side to generate and reconstruct SMS sequence. Can you merge it at some point ?

A missing feature will be have have standard 8/16/32 coils that are close to something real ? Or at least to share in the doc a good snippets to generate them

Commit:
CompatHelper: bump compat for Interpolations to 0.16 for package KomaMRIBase, (keep existing compat)

Commit:
CompatHelper: bump compat for Interpolations to 0.16 for package KomaMRIPlots, (keep existing compat)

Stockless committed:
update Bloch_RF_accuracy test to use DiffEq instead of pre-calculated reference values.

Stockless committed:
delete debugging test

Stockless committed:
admission of ArbitraryRFCoils simulation without interpolation (only coils_coords at phantom positions)

Stockless committed:
Moved coils_sens interpolation to SimulatorCore

Stockless committed:
Moved coils_sens interpolation to SimulatorCore

Stockless committed:
Fix sys::Scanner -> sys::Scanner{T}

Stockless committed:
Fix sys::Scanner -> sys::Scanner{T}

Commit:
CompatHelper: bump compat for AcceleratedKernels to 0.4 for package KomaMRICore, (keep existing compat)

Stockless committed:
Redaction corrections

Stockless committed:
julia code block correction

Stockless committed:
Generate preview

Stockless committed:
Corrected redaction according to review

Stockless committed:
Update Phantom use on examples

Stockless committed:
Update Phantom use on tests

Stockless committed:
Update Phantom use on definitions (UI,Base,Core)

Stockless committed:
add tags to PR

Stockless committed:
Improve test-komamricore image resolution

pvillacorta committed:
Add sequence scripts folder and new  `lit-07-RRVariability.jl` tutorial

pvillacorta committed:
Remove CUDA import from `examples/3.tutorials/lit-07-RRVariability.jl`

pvillacorta committed:
Rename bSSFP_cine.jl and fix include functions

pvillacorta committed:
Update docs/Project.toml to include Plots and Printf

pvillacorta committed:
Utils file redistribution and  correct bugs that difficulted docs build

pvillacorta committed:
Make struct definitions mutable for TimeCurve, AllSpins, SpinRange, FlowPath, Path, HeartBeat, Rotate, and Translate

pvillacorta committed:
Finish lit-07 tutorial and redistribute /utils files

pvillacorta committed:
Fix commands tagged with `#nb` and `#jl`

pvillacorta committed:
Apply suggested changes, add phantom plot, resize figures

pvillacorta committed:
Remove `mutable` from all the motion-related structs

pvillacorta committed:
Correct lit-07-RRVariability.jl to modify non-mutable structs

Stockless committed:
Improved redaction based on review

pvillacorta committed:
Commit of the following:
- `execute=true` in `Literate.markdown` function
- Add Suppressor and DisplayAs packages
- Adapt all examples to `execute=true` option

Stockless committed:
Interpolation Incorporation

Stockless committed:
Interpolation Incorporation

Stockless committed:
News 16.Jun.2025

Stockless committed:
add space after images

Stockless committed:
Remove unnecesary itemization

cncastillo committed:
Update README.md

Stockless committed:
Remove CSS specification to generate space after image

cncastillo committed:
Add files via upload

cncastillo committed:
Delete assets/OSI2Logo.svg

cncastillo committed:
Add files via upload

cncastillo committed:
Rename DIN SPEC 3105_ OSI² CAB_ 1.svg to DIN_SPEC_3105_OSI2_CAB_1.svg

cncastillo committed:
Added OSI2 badge

cncastillo committed:
Added OSI2 badge properly

cncastillo committed:
Delete assets/DIN_SPEC_3105_OSI2_CAB_1.svg

Stockless committed:
fix tabulation errors

Stockless committed:
add coil_sens view

Stockless committed:
test new function instead of macro

Stockless committed:
using blink outside linux condition

KookiesNKareem committed:
Fixed inconsistencies in header capitalization, found and fixed various typos and unclear language

KookiesNKareem committed:
Fixed more capitalization inconsistencies and typos. Clarified the sync changes step in "How To Commit"

Stockless committed:
"static if" and rm innecesary export

Stockless committed:
include macos in temporary fix

Stockless committed:
add sleep after window generation

Stockless committed:
initialize interpolated_coil_sens with similar()

Stockless committed:
temporary change

cncastillo committed:
Always return window.

cncastillo committed:
Just import Blink.

cncastillo committed:
Always return window.

Stockless committed:
test using Blink.Window directly

Stockless committed:
window without arguments

Stockless committed:
replace setup-julia@v2 with latest

Stockless committed:
Add lynx for debugging electron

Stockless committed:
test sleep on ExportUIFunctions.jl

Stockless committed:
add actions to test

Stockless committed:
sleep closer to window generation

Stockless committed:
skip macos ui tests

Stockless committed:
fix if statement

Stockless committed:
using Blink outside condition

Stockless committed:
restablish commented sections of tests

Stockless committed:
comment unnecesary "include"s

Stockless committed:
final changes to temporary fix

Stockless committed:
simulate with interpolation (reconstruction missing)

cncastillo committed:
Update lit-01-FID.jl

Stockless committed:
re-organize coils definition example

Stockless committed:
Bloch_RF_accuracy NRMSE improve to be > 0.1

Stockless committed:
rm println

Stockless committed:
modified duration = dur(seq)

cncastillo committed:
Reformat test

Stockless committed:
change coil_sens_recon

Stockless committed:
rebase remaining fixes

Stockless committed:
Merge branch 'master' into scanner-develop

Stockless committed:
complete rebase

cncastillo commented on KomaMRIBase/src/datatypes/Scanner.jl:
```diff
@@ -1,3 +1,48 @@
+
+# Gradients
+abstract type Gradients{T} end
+@with_kw mutable struct LinearXYZGradients{T} <: Gradients{T}
+    Gx::AbstractVector{T} = zeros(T, 1)
+    Gy::AbstractVector{T} = zeros(T, 1)
+    Gz::AbstractVector{T} = zeros(T, 1)
+end
```
This should have anything inside.

cncastillo commented on KomaMRIBase/src/datatypes/Scanner.jl:
```diff
@@ -28,19 +73,73 @@ julia> sys = Scanner()
+
+function acquire_signal!(sig, obj, rf_coils::ArbitraryRFCoils, Mxy)
+    interpolated_coil_sens = similar(rf_coils.coil_sens, length(obj.x), size(rf_coils.coil_sens, 4))
+    coords = collect(zip(obj.x, obj.y, obj.z)) 
```
Remove `collect` this is very bad practice. Also `zip`ing the `obj.x`, etc. is not good.

cncastillo commented on KomaMRIBase/src/datatypes/Scanner.jl:
```diff
@@ -28,19 +73,73 @@ julia> sys = Scanner()
+    coords = collect(zip(obj.x, obj.y, obj.z)) 
+    for i in 1:size(rf_coils.coil_sens, 4)
+        itp = GriddedInterpolation((rf_coils.x, rf_coils.y, rf_coils.z), rf_coils.coil_sens[:,:,:,i], Gridded(Linear()))
+        eitp = extrapolate(itp, 0.0 + 0.0im)
+        interpolated_coil_sens[:,i] = map(p -> eitp(p...), coords)
```
Modify to follow conventions from Interpolations.jl documentation.

cncastillo commented on KomaMRIBase/src/datatypes/Scanner.jl:
```diff
@@ -28,19 +73,73 @@ julia> sys = Scanner()
+        eitp = extrapolate(itp, 0.0 + 0.0im)
+        interpolated_coil_sens[:,i] = map(p -> eitp(p...), coords)
+    end
+    for i in 1:size(interpolated_coil_sens, 2)
```
this for loop is the same as the previous

cncastillo commented on KomaMRIBase/test/runtests.jl:
```diff
@@ -737,10 +737,23 @@ end
     ADC_Δt, seq_Δt, GR_Δt, RF_Δt = 2e-6, 1e-5, 1e-5, 1e-6
     RF_ring_down_T, RF_dead_time_T, ADC_dead_time_T = 20e-6, 100e-6, 10e-6
-    sys = Scanner(B0, B1, Gmax, Smax, ADC_Δt, seq_Δt, GR_Δt, RF_Δt, RF_ring_down_T, RF_dead_time_T, ADC_dead_time_T)
+    limits = HardwareLimits(
+        B0 = B0, 
+        B1 = B1, 
+        Gmax = Gmax, 
+        Smax = Smax, 
+        ADC_Δt = ADC_Δt, 
+        seq_Δt = seq_Δt, 
+        GR_Δt = GR_Δt, 
+        RF_Δt = RF_Δt, 
+        RF_ring_down_T = RF_ring_down_T, 
+        RF_dead_time_T = RF_dead_time_T, 
+        ADC_dead_time_T = ADC_dead_time_T
+    )
```
Why did you modify this?

Stockless replied:
previous test had this definition of scanner:
    `sys = Scanner(B0, B1, Gmax, Smax, ADC_Δt, seq_Δt, GR_Δt, RF_Δt, RF_ring_down_T, RF_dead_time_T, ADC_dead_time_T)`
This is not longer the way of defining this values.

cncastillo replied:
But why are you using `param=param`, you can use the param once.

cncastillo commented on examples/3.tutorials/lit-02-SmallTipApproximation.jl:
```diff
@@ -2,9 +2,9 @@
-using KomaMRI, Suppressor #hide
-sys = Scanner() #hide
-sys.Smax = 50; #hide
+using KomaMRI # hide
+sys = Scanner() # hide
+sys.Smax = 50.0 # hide
```
Do not mofiy

cncastillo committed:
Update README.md [skip CI]

cncastillo committed:
Update README.md [skip CI]

Stockless committed:
3D interpolation

Stockless committed:
Update runbenchmarks.jl

Stockless committed:
Update runbenchmarks.jl

Stockless committed:
update enviroments for buildkite test

Stockless committed:
update enviroments for buildkite test

cncastillo committed:
Fix environment in which tests are run

cncastillo committed:
Fix environment in which tests are run

cncastillo committed:
Test environment change

cncastillo committed:
Test environment change

cncastillo committed:
Update test environment setup (again)

cncastillo committed:
Update test environment setup (again)

cncastillo committed:
Last update to test environment setup

cncastillo committed:
Last update to test environment setup

cncastillo committed:
Add GPU backend to test  environment

cncastillo committed:
Add GPU backend to test  environment

cncastillo committed:
Fix GPU tests re-resolving test environment

cncastillo committed:
Fix GPU tests re-resolving test environment

cncastillo committed:
KomaMRIBase dev'ed in project environment

cncastillo committed:
KomaMRIBase dev'ed in project environment

cncastillo committed:
Dev packages to test environment

cncastillo committed:
Dev packages to test environment

cncastillo committed:
Only use tests environment for GPU tests

cncastillo committed:
Only use tests environment for GPU tests

cncastillo committed:
Test environment setup different for 1.10

cncastillo committed:
Test environment setup different for 1.10

cncastillo committed:
Fix oneAPI bool indexing issues 

This is generating too many issues and is only happening in `BlochSimple` which is not performance critical. Therefore, I am replacing the `B[B .== 0] .= eps(T)` in [line 92](https://github.com/JuliaHealth/KomaMRI.jl/blob/master/KomaMRICore/src/simulation/SimMethods/BlochSimple/BlochSimple.jl#L92C9-L92C29) by `B .+= (B .== 0) .* eps(T)` which shouldn't have any problems, but it should be a little slower.

cncastillo committed:
Fixes oneAPI indexing issues.

cncastillo committed:
Add tags to Buildkite codecov process

cncastillo committed:
Update README.md [skip CI]

cncastillo committed:
Update README.md [skip CI]

cncastillo committed:
Flags added to official julia-coverage

Stockless committed:
interpolate 2D/3D cases

jonbmartin commented:
In my own testing this has worked nicely for a noncartesian sim. Would love to see fully incorporated. Cool development.

Stockless committed:
Replace coil_tange example definition

Stockless committed:
`sig[1]` fill correction

Stockless committed:
Merge branch 'master' into scanner-develop

Stockless committed:
aquire_signal on GPU

Stockless committed:
phantom with complex rho

Stockless committed:
Merge branch 'master' into scanner-develop

aTrotier commented:
Can I help on this PR ? :)
